### PR TITLE
EOS-13990: NFS ADDB: Tracepointsfor all ATTR operations

### PR DESCRIPTION
**Pull request by vimalraghwani20**
**Problem Statement**
EOS-13990: NFS ADDB: Tracepointsfor all ATTR operations
EOS-10522: NFS ADDB: Decoders for all file operations

**Problem Description**
Implement tracepoints and decoder for the attr handlers.

**Solution Overview**
Following changes has been done:

- added wrapper over kvsfs_getattrs  using kvsfs_perf_op_getattrs
- Added traces for init and finish inside kvsfs_getattrs
- Added traces for attributes like return code   
- added wrapper over kvsfs_setattrs   using kvsfs_perf_op_setattrs
- Added traces for init and finish inside kvsfs_setattrs
- Added traces for attributes like return code
- Added all enums in the decoder.

**Unit Test Cases**
`
sudo m0addb2dump -p /usr/lib64/libcortx-utils.so -- /addb_22028/o/100000000000000\:2 | grep fsuser > addb_dump_op_decode_4	

-bash-4.2$ cat addb_dump_op_decode_4 | grep "fsal_getattrs"
* 2020-10-08-01:13:36.571926218 fsuser_state     fsal_getattrs, state,   e,          init
* 2020-10-08-01:13:36.576515604 fsuser_attribute fsal_getattrs, attribute, e,        get_attribute_result_major_code, ?               0?
* 2020-10-08-01:13:36.576516264 fsuser_attribute fsal_getattrs, attribute, e,        get_attribute_result_minor_code, ?               0?
* 2020-10-08-01:13:36.576516774 fsuser_state     fsal_getattrs, state,   e,          finish
* 2020-10-08-01:14:01.444814040 fsuser_state     fsal_getattrs, state,   36,         init
* 2020-10-08-01:14:01.448805051 fsuser_attribute fsal_getattrs, attribute, 36,       get_attribute_result_major_code, ?               0?
* 2020-10-08-01:14:01.448805701 fsuser_attribute fsal_getattrs, attribute, 36,       get_attribute_result_minor_code, ?               0?
* 2020-10-08-01:14:01.448806226 fsuser_state     fsal_getattrs, state,   36,         finish
* 2020-10-08-01:14:16.542089191 fsuser_state     fsal_getattrs, state,   c3,         init
* 2020-10-08-01:14:16.546121912 fsuser_attribute fsal_getattrs, attribute, c3,       get_attribute_result_major_code, ?               0?
* 2020-10-08-01:14:16.546122558 fsuser_attribute fsal_getattrs, attribute, c3,       get_attribute_result_minor_code, ?               0?
* 2020-10-08-01:14:16.546123094 fsuser_state     fsal_getattrs, state,   c3,         finish
* 2020-10-08-01:14:29.730222059 fsuser_state     fsal_getattrs, state,   17c,        init
* 2020-10-08-01:14:29.734102240 fsuser_attribute fsal_getattrs, attribute, 17c,      get_attribute_result_major_code, ?               0?
* 2020-10-08-01:14:29.734102870 fsuser_attribute fsal_getattrs, attribute, 17c,      get_attribute_result_minor_code, ?               0?
* 2020-10-08-01:14:29.734103387 fsuser_state     fsal_getattrs, state,   17c,        finish

-bash-4.2$ cat addb_dump_op_decode_4 | grep "fsal_setattrs"
* 2020-10-08-01:14:34.079656624 fsuser_state     fsal_setattrs, state,   566,        init
* 2020-10-08-01:14:34.082418414 fsuser_attribute fsal_setattrs, attribute, 566,      set_attribute_result_major_code, ?               0?
* 2020-10-08-01:14:34.082419134 fsuser_attribute fsal_setattrs, attribute, 566,      set_attribute_result_minor_code, ?               0?
* 2020-10-08-01:14:34.082419712 fsuser_state     fsal_setattrs, state,   566,        finish
* 2020-10-08-01:14:33.909490008 fsuser_state     fsal_setattrs, state,   492,        init
* 2020-10-08-01:14:33.912997100 fsuser_attribute fsal_setattrs, attribute, 492,      set_attribute_result_major_code, ?               0?
* 2020-10-08-01:14:33.912997868 fsuser_attribute fsal_setattrs, attribute, 492,      set_attribute_result_minor_code, ?               0?
* 2020-10-08-01:14:33.912998408 fsuser_state     fsal_setattrs, state,   492,        finish
* 2020-10-08-01:14:55.922933569 fsuser_state     fsal_setattrs, state,   7eb,        init
* 2020-10-08-01:14:55.926658809 fsuser_attribute fsal_setattrs, attribute, 7eb,      set_attribute_result_major_code, ?               0?
* 2020-10-08-01:14:55.926660101 fsuser_attribute fsal_setattrs, attribute, 7eb,      set_attribute_result_minor_code, ?               0?
* 2020-10-08-01:14:55.926660668 fsuser_state     fsal_setattrs, state,   7eb,        finish
* 2020-10-08-01:14:52.125471046 fsuser_state     fsal_setattrs, state,   76e,        init
* 2020-10-08-01:14:52.128594395 fsuser_attribute fsal_setattrs, attribute, 76e,      set_attribute_result_major_code, ?               0?
* 2020-10-08-01:14:52.128595145 fsuser_attribute fsal_setattrs, attribute, 76e,      set_attribute_result_minor_code, ?               0?
* 2020-10-08-01:14:52.128595686 fsuser_state     fsal_setattrs, state,   76e,        finish

`

### CODE REVIEW

**Comment by AtitaShirwaikar** [comment]
Noticed that you do not have tracepoints added at CFS, NSAL layer. Were you able to leverage tracepoints added by @pratyush-seagate for READ operation or will you be raising a PR for that?

File: src/FSAL/FSAL_CORTXFS/handle.c [hunk 1]
```diff
@@ -1236,9 +1267,23 @@ static fsal_status_t kvsfs_getattrs(struct fsal_obj_handle *obj_hdl,
 
 out:
 	T_EXIT0(result.major);
+    perfc_trace_attr(PEA_GETATTR_RES_MAJ, result.major);
```
Comment (AtitaShirwaikar) [inline src/FSAL/FSAL_CORTXFS/handle.c:1270]: Would it be prudent to add tracepoints for any other attributes like members of fsal_obj_handle?

File: src/FSAL/FSAL_CORTXFS/handle.c [hunk 2]
```diff
@@ -1406,9 +1452,24 @@ fsal_status_t kvsfs_setattrs(struct fsal_obj_handle *obj_hdl,
 
 out:
 	T_EXIT0(result.major);
+    perfc_trace_attr(PEA_SETATTR_RES_MAJ, result.major);
```
Comment (AtitaShirwaikar) [inline src/FSAL/FSAL_CORTXFS/handle.c:1455]: Here too.Would it be prudent to add tracepoints for any other attributes like members of fsal_obj_handle?

**Comment by pratyush-seagate** [comment]
These changes are ok, but can you please also add traces in the interim NSAL and/or DSAL layer functions present the code path for set/get attrs?

File: src/FSAL/FSAL_CORTXFS/handle.c (hunk 1 above)
Comment (vimalraghwani20) [reply src/FSAL/FSAL_CORTXFS/handle.c:1270]: As of now, ADDB traces for integer variable can be logged.

File: src/FSAL/FSAL_CORTXFS/handle.c (hunk 2 above)
Comment (vimalraghwani20) [reply src/FSAL/FSAL_CORTXFS/handle.c:1455]: As of now, ADDB traces for integer variable can be logged.

**Review by pratyush-seagate — APPROVED** [review]
LGTM, +1

File: src/FSAL/FSAL_CORTXFS/handle.c [hunk 3]
```diff
@@ -1319,10 +1364,10 @@ static fsal_status_t kvsfs_ftruncate(struct fsal_obj_handle *obj,
 				     struct stat *new_stat, int new_stat_flags);
 
 /* FSAL.setattrs2 */
-fsal_status_t kvsfs_setattrs(struct fsal_obj_handle *obj_hdl,
-			     bool bypass,
-			     struct state_t *state,
-			     struct attrlist *attrs)
+static fsal_status_t kvsfs_setattrs(struct fsal_obj_handle *obj_hdl,
```
Comment (sachinpunadikar) [inline src/FSAL/FSAL_CORTXFS/handle.c:1367]: Shouldn't this be "static inline" ? @pratyush-seagate what you say ?

Comment (pratyush-seagate) [reply src/FSAL/FSAL_CORTXFS/handle.c:1367]: Yes, better to add inline.

Comment (pratyush-seagate) [reply src/FSAL/FSAL_CORTXFS/handle.c:1367]: But inline keyword is just a suggestion to compiler, which it can ignore.

Comment (sachinpunadikar) [reply src/FSAL/FSAL_CORTXFS/handle.c:1367]: Agreed, but lets take up the advantage wherever compiler can.

Comment (vimalraghwani20) [reply src/FSAL/FSAL_CORTXFS/handle.c]: fixed.

**Review by sachinpunadikar — APPROVED** [review]
+1

**Pull request merged**